src/lightning/data/streaming/dataset_optimizer.py

```diff
@@ -790,6 +819,16 @@ def _upload_index(self, cache_dir: str, num_nodes: int, node_rank: Optional[int]
                 node_index_filepath = os.path.join(cache_dir, os.path.basename(remote_filepath))
                 if obj.scheme == "s3":
                     obj = parse.urlparse(remote_filepath)
+                    # Wait for the index to exist on s3.
```

Comment:
I would refactor this code so that it can be more easily tested. Right now, you can't test it well because there is a while loop. Or maybe you can mock `s3.head_object` with a side-effect that raises an exception.

Reply:
Done.